service/storage_service.cc

```diff
@@ -1777,6 +1777,8 @@ future<> storage_service::join_token_ring(cdc::generation_service& cdc_gen_servi
 
         co_await update_topology_with_local_metadata(*raft_server);
 
+        co_await update_fence_version(co_await _sys_ks.local().get_topology_fence_version());
```

Comment:
> Won't this race with update_fence_version coming from coordinator commands, which might be arriving at this point?

I was under impression that we don't accept external communication until that point

> Also on restart I think we should do this earlier - because IIRC at this point we might already be performing data path ops.

Indeed, `messaging_service::start_listen` is called before `storage_service::join_cluster`, so we should call this somewhere before.

Reply:
> > I was under impression that we don't accept external communication until that point
> 
> How would we join group 0 and perform a read barrier?

May patch will create group0 server before any message is received to solve the race where messages that need group0 started to arrive before it was created. Doing anything through this server will be possible only after messaging service starts listening.